File: 12-27-2023-suffix-sort.py
def merge(a1, a2):
    mergedArrs = []
    i = 0
    j = 0
    while i < len(a1) or j < len(a2):
        if i == len(a1):
            mergedArrs.append(a2[j])
            j += 1
        elif j == len(a2):
            mergedArrs.append(a1[i])
            i += 1
        elif a1[i] <= a2[j]:
            mergedArrs.append(a1[i])
            i += 1
        else:
            mergedArrs.append(a2[j])
            j += 1
    return mergedArrs


def sortSuffixArr(a):
    if len(a) <= 1:
        return a
    else:
        mid = len(a) // 2
        return merge(sortSuffixArr(a[0:mid]), sortSuffixArr(a[mid:]))


def suffixSort(s, k):
    suffixes = []

    for i in range(len(s)):
        suffixes.append(s[i:])

    sortedSuffixes = sortSuffixArr(suffixes)

    return sortedSuffixes[k - 1]


print(suffixSort("aacb", 3))

File: Fun Practice/12-27-2023-suffix-sort.js
function merge(arr1, arr2) {
  let mergedArrs = [];
  let i = 0;
  let j = 0;
  while (i < arr1.length || j < arr2.length) {
    if (i === arr1.length) {
      mergedArrs.push(arr2[j]);
      j++;
    } else if (j === arr2.length) {
      mergedArrs.push(arr1[i]);
      i++;
    } else if (arr1[i] <= arr2[j]) {
      mergedArrs.push(arr1[i]);
      i++;
    } else {
      mergedArrs.push(arr2[j]);
      j++;
    }
  }
  return mergedArrs;
}

function sortSuffixArr(arr) {
  if (arr.length <= 1) {
    return arr;
  } else {
    let mid = Math.floor(arr.length / 2);
    return merge(
      sortSuffixArr(arr.slice(0, mid)),
      sortSuffixArr(arr.slice(mid))
    );
  }
}

function suffixSort(str, k) {
  let suffixes = [];

  for (let i = 0; i < str.length; i++) {
    suffixes.push(str.slice(i, str.length));
  }

  let sortedSuffixes = sortSuffixArr(suffixes);

  return sortedSuffixes;
}

console.log(suffixSort("aasdgergaseg", 3));
